atd-vze/src/Components/GridExportData.js

```diff
@@ -0,0 +1,166 @@
+  let [getExport, { loading, data }] = useLazyQuery(
+    query.queryCSV(columnsToExport)
+  );
+
+  const toggleModal = () => setIsModalOpen(!isModalOpen);
+
+  // Open modal and execute initial query
+  const toggleModalAndExport = () => {
+    setIsModalOpen(!isModalOpen);
+    getExport();
+  };
+
+  // Set query limit from inputs and then execute query
+  const setExportLimit = event => {
+    // Limit set to 1 so that query doesn't include all records unintentionally
+    if (event.target.id === "csv-number-input") {
+      query.limit = event.target.value !== "" ? event.target.value : 1;
+    } else if ((event.target.id = "csv-checkbox-input")) {
+      query.limit = event.target.checked ? totalRecords : 1;
+    }
+    getExport();
+  };
+
+  /**
+   * Returns an array of objects (each object is a row and each key of that object is a column in the export file)
+   * @param {Array} Data returned from DB with nested data structures
+   * @returns {array}
+   */
+  const formatExportData = data => {
+    // Moves nested data to top level object (CSVLink uses each top level key as a column header)
+    // Handles:
+    // 1. Nested objects
+    // 2. Nested arrays of objects
+    const newData = data.map(row => {
+      // Look through each row for nested objects or arrays and move all to top level
+      Object.entries(row).forEach(([key, value]) => {
+        // Remove __typename from export (contains table name which is already in filename)
+        if (key === "__typename") {
+          delete row["__typename"];
+        } else if (Array.isArray(value)) {
+          // If value is array, recursive call and handle objects in array
+          value = formatExportData(value);
+          value.forEach(object => {
+            Object.entries(object).forEach(([key, value]) => {
+              if (row[key]) {
+                // If top level already has this key, concat
+                row[key] = `${row[key]}, ${value}`;
+              } else {
+                // Else use spread to add to top level
+                row = { ...row, ...object };
+              }
+            });
+            // Delete nested data after added to top level
+            delete row[key];
+          });
+        } else if (typeof value === "object" && value !== null) {
+          // If value is object, remove __typename and move to top level, then delete
+          "__typename" in value && delete value["__typename"];
+          row = { ...row, ...value };
+          delete row[key];
+        }
+      });
+      return row;
+    });
+    return newData;
+  };
+
+  return (
+    <>
+      <StyledSaveLink>
+        <i
+          className="fa fa-save fa-2x ml-2 mt-1"
+          onClick={toggleModalAndExport}
+        />
+      </StyledSaveLink>
+      <Modal isOpen={isModalOpen} toggle={toggleModal} className={"modal-sm "}>
+        <ModalHeader toggle={toggleModal}>
+          Export to .csv ({query.limit} rows)
+        </ModalHeader>
+        <ModalBody>
```

Comment:
Thinking here we add a conditional `<Alert />` component if the `totalRecords` > 4k records. "For larger downloads, please expect a delay while the CSV file is generated. This may take multiple minutes."

Reply:
This is a great idea because it is a really long wait for the larger exports. Sergio brought this up as well, and I could see the spinner going for minutes on end leading to some confusion. I'll add this before merge.